subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTaskFactoryTest.groovy

```diff
@@ -572,6 +573,50 @@ class AnnotationProcessingTaskFactoryTest extends AbstractProjectBuilderSpec {
         TaskWithBooleanInput                      | "inputValue"      | true                     | true           // https://issues.gradle.org/Browse/GRADLE-2815
     }
 
+    @Unroll
+    def "registers properties #allTaskProperties on #type.simpleName"() {
+        given:
+        def task = (value == null) ? expectTaskCreated(type) : expectTaskCreated(type, value as Object[])
+
+        when:
+        def taskProperties = DefaultTaskProperties.resolve(propertyWalker, project.fileResolver, task)
+
+        then:
+        taskProperties.inputPropertyValues.create().keySet() == inputs as Set
+        taskProperties.inputFileProperties*.propertyName as Set == inputFiles as Set
+        taskProperties.outputFileProperties*.propertyName as Set == outputFiles as Set
```

Comment:
Should we also test `@Destroyable` and `@LocalState` discovery?

Reply:
I thin we should compare these as `List`s to ensure that the ordering is how we expect it, too.

Reply:
What kind of order would we guarantee here? Especially since `inputPropertyValues` is a `HashMap`, I would not expect a certain order there. For input files and output files we already return sorted sets. Would you expect to check the ordering of the sorted sets here?

Reply:
I added some tests for `@Destroyable` and `@LocalState` discovery.